src/celeritas/optical/OpticalTrackData.hh

```diff
@@ -0,0 +1,162 @@
+#include "celeritas/random/RngData.hh"
+#include "celeritas/track/SimData.hh"
+#include "celeritas/track/TrackInitData.hh"
+
+#include "Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+// XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX
+// IMPLEMENT ME!
+
+template<Ownership W, MemSpace M>
+struct OpticalPhysicsParamsData
+{
+    explicit CELER_FUNCTION operator bool() const { return false; }
+};
+template<Ownership W, MemSpace M>
+struct OpticalPhysicsStateData
+{
+};
+template<MemSpace M>
+void resize(OpticalPhysicsStateData<Ownership::value, M>*,
+            HostCRef<OpticalPhysicsParamsData> const&,
+            size_type)
+{
+    CELER_NOT_IMPLEMENTED("optical physics state");
+}
+
+// XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX  XXX
+//---------------------------------------------------------------------------//
+/*!
+ * Memspace-independent core variables.
+ */
+struct OpticalScalars
+{
+    ActionId boundary_action;
+
+    StreamId::size_type max_streams{0};
+    OpticalMaterialId::size_type num_materials{0};
+
+    //! True if assigned and valid
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return boundary_action && max_streams > 0 && num_materials > 0;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Immutable problem data.
+ */
+template<Ownership W, MemSpace M>
+struct OpticalParamsData
+{
+    template<class T>
+    using VolumeItems = celeritas::Collection<T, W, M, VolumeId>;
+
+    GeoParamsData<W, M> geometry;
+    VolumeItems<OpticalMaterialId> materials;
+    OpticalPhysicsParamsData<W, M> physics;
+    RngParamsData<W, M> rng;
+    TrackInitParamsData<W, M> init;  // TODO: don't need max events
+
+    OpticalScalars scalars;
+
+    //! True if all params are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return geometry && !materials.empty() && physics && rng && init
+               && scalars;
+    }
+
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    OpticalParamsData& operator=(OpticalParamsData<W2, M2> const& other)
+    {
+        CELER_EXPECT(other);
+        geometry = other.geometry;
+        materials = other.materials;
+        physics = other.physics;
+        rng = other.rng;
+        init = other.init;
+        scalars = other.scalars;
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Thread-local state data.
+ */
+template<Ownership W, MemSpace M>
+struct OpticalStateData
+{
+    template<class T>
+    using Items = StateCollection<T, W, M>;
+
+    GeoStateData<W, M> geometry;
+    Items<OpticalMaterialId> materials;
+    OpticalPhysicsStateData<W, M> physics;
+    RngStateData<W, M> rng;
+    SimStateData<W, M> sim;  // TODO: has a few things we don't need
```

Comment:
Which sim data do we not need? I guess `num_steps`, `num_looping_steps`, `along_step_action`... do we still need the track/parent/event IDs?